pkg/apis/pipeline/v1/resultref.go

```diff
@@ -138,24 +149,44 @@ func stripVarSubExpression(expression string) string {
 // - Input: tasks.myTask.results.resultName.foo.bar
 // - Output: "", "", nil, "", error
 // TODO: may use regex for each type to handle possible reference formats
-func parseExpression(substitutionExpression string) (string, string, *int, string, error) {
-	if looksLikeResultRef(substitutionExpression) {
+func parseExpression(substitutionExpression string) (string, string, *int, string, ParamType, error) {
+	if looksLikeResultRef(substitutionExpression) || looksLikeStepResultRef(substitutionExpression) {
 		subExpressions := strings.Split(substitutionExpression, ".")
 		// For string result: tasks.<taskName>.results.<stringResultName>
+		// For string step result: steps.<stepName>.results.<stringResultName>
 		// For array result: tasks.<taskName>.results.<arrayResultName>[index]
+		// For array step result: steps.<stepName>.results.<arrayResultName>[index]
 		if len(subExpressions) == 4 {
 			resultName, stringIdx := ParseResultName(subExpressions[3])
-			if stringIdx != "" && stringIdx != "*" {
+			if stringIdx != "" {
+				if stringIdx == "*" {
+					return subExpressions[1], resultName, nil, "", ParamTypeArray, nil
+				}
 				intIdx, _ := strconv.Atoi(stringIdx)
-				return subExpressions[1], resultName, &intIdx, "", nil
+				return subExpressions[1], resultName, &intIdx, "", ParamTypeArray, nil
 			}
-			return subExpressions[1], resultName, nil, "", nil
+			return subExpressions[1], resultName, nil, "", ParamTypeString, nil
 		} else if len(subExpressions) == 5 {
 			// For object type result: tasks.<taskName>.results.<objectResultName>.<individualAttribute>
-			return subExpressions[1], subExpressions[3], nil, subExpressions[4], nil
+			// For object type step result: steps.<stepName>.results.<objectResultName>.<individualAttribute>
+			return subExpressions[1], subExpressions[3], nil, subExpressions[4], ParamTypeObject, nil
 		}
 	}
-	return "", "", nil, "", fmt.Errorf("must be one of the form 1). %q; 2). %q", resultExpressionFormat, objectResultExpressionFormat)
+	return "", "", nil, "", ParamTypeString, fmt.Errorf("must be one of the form 1). %q; 2). %q; 3). %q; 4). %q", resultExpressionFormat, objectResultExpressionFormat, stepResultExpressionFormat, objectStepResultExpressionFormat)
+}
+
+func parseTaskExpression(substitutionExpression string) (string, string, *int, string, ParamType, error) {
+	if looksLikeResultRef(substitutionExpression) {
+		return parseExpression(substitutionExpression)
+	}
+	return "", "", nil, "", ParamTypeString, fmt.Errorf("must be one of the form 1). %q; 2). %q", resultExpressionFormat, objectResultExpressionFormat)
+}
+
+func ParseStepExpression(substitutionExpression string) (string, string, *int, string, ParamType, error) {
```

Comment:
Add godoc.

There's enough return values here it probably makes sense to break this out into its own struct with field names + godoc.

Reply:
fixed.